databroker/broker.py

```diff
@@ -562,3 +566,130 @@ def _munge_time(t, timezone):
     """
     t = datetime.fromtimestamp(t)
     return timezone.localize(t).replace(microsecond=0).isoformat()
+
+
+known_special_keys = {
+    # The duration of the scan
+    'duration': lambda header: (datetime.fromtimestamp(header.stop.time) -
+                                datetime.fromtimestamp(header.start.time)),
+    # The number of events per event descriptor
+    'num_events': lambda header: [
+        len(list(mdsc.get_events_generator(descriptor, False)))
+        for descriptor in header.descriptors],
+}
+
+def _get_uid(key, document):
+    try:
+        # see if we have a hyphen
+        key, num = key.split('-')
+        num = int(num)
+    except ValueError:
+        # Nope! Must be asking for the whole uid
+        num = None
+    return document.uid[:num]
+
+
+@singledispatch
+def _get_value(h, key, default_value):
+    """
+    Get the value specified by `key` from the Header `h`. Give the value
+    of `default_value` for keys that are not in `h`
+
+    Parameters
+    ----------
+    h : databroker.core.Header
+        A single header to try and get a key from
+    key : str
+        The key to find a value for in `h`. See the `*keys` section of the
+        `databroker.broker.summarize` docstring
+    default_value : str
+        The default value for which to fill missing keys
```

Comment:
If you have multiple headers, some that have keys and some that don't, you will want a default value